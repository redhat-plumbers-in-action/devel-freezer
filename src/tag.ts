import { Context } from 'probot';

import { events } from './events';

export class Tag {
  private _latest: string | undefined;

  constructor(tag: string | undefined) {
    this._latest = tag;
  }

  get latest() {
    return this._latest;
  }

  isFreezed(tagPolicy: string[]) {
    if (this.latest === undefined) false;

    return tagPolicy.some(regex =>
      new RegExp(regex).test(this.latest as string)
    );
  }

  static async getLatestTag(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const tags = (await context.octokit.rest.repos.listTags(context.repo()))
      .data;

    if (!tags || tags.length <= 0) return undefined;

    return tags.shift()?.name;
  }
}
