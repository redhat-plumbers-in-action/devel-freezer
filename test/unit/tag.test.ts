import { describe, it, expect, beforeEach, test } from 'vitest';

import { tagContextFixture, ITagTestContext } from './fixtures/tag.fixture';

describe('Tag Object', () => {
  beforeEach<ITagTestContext>(context => {
    context.tags = tagContextFixture.tags;
    context.tagPolicy = tagContextFixture.tagPolicy;
    context.invalid = tagContextFixture.invalid;
  });

  it<ITagTestContext>('can be instantiated', context =>
    context.tags.map(tagsItem => expect(tagsItem).toBeDefined()));

  it.todo<ITagTestContext>(
    'is invalid' /*, async context => {
    context.invalid.map(async item =>
      expect(Metadata.validate(item)).resolves.toMatchSnapshot()
    );
  }*/
  );

  test<ITagTestContext>('get latest()', context =>
    context.tags.map(tagsItem => expect(tagsItem.latest).toMatchSnapshot()));

  test<ITagTestContext>('isFreezed()', context =>
    context.tags.map(tagsItem =>
      context.tagPolicy.map(tagPolicyItem =>
        expect(tagsItem.isFreezed(tagPolicyItem)).toMatchSnapshot()
      )
    ));
});
