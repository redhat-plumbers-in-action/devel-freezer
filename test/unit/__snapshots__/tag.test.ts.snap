// Vitest Snapshot v1

exports[`Tag Object > get latest() 1`] = `"v0.0.1"`;

exports[`Tag Object > get latest() 2`] = `"v1"`;

exports[`Tag Object > get latest() 3`] = `"latest"`;

exports[`Tag Object > get latest() 4`] = `"v1-rc1"`;

exports[`Tag Object > get latest() 5`] = `"v1-beta"`;

exports[`Tag Object > get latest() 6`] = `"alpha"`;

exports[`Tag Object > get latest() 7`] = `"beta"`;

exports[`Tag Object > get latest() 8`] = `"rc1"`;

exports[`Tag Object > isFreezed() 1`] = `false`;

exports[`Tag Object > isFreezed() 2`] = `false`;

exports[`Tag Object > isFreezed() 3`] = `false`;

exports[`Tag Object > isFreezed() 4`] = `false`;

exports[`Tag Object > isFreezed() 5`] = `false`;

exports[`Tag Object > isFreezed() 6`] = `false`;

exports[`Tag Object > isFreezed() 7`] = `false`;

exports[`Tag Object > isFreezed() 8`] = `false`;

exports[`Tag Object > isFreezed() 9`] = `true`;

exports[`Tag Object > isFreezed() 10`] = `false`;

exports[`Tag Object > isFreezed() 11`] = `false`;

exports[`Tag Object > isFreezed() 12`] = `false`;

exports[`Tag Object > isFreezed() 13`] = `false`;

exports[`Tag Object > isFreezed() 14`] = `true`;

exports[`Tag Object > isFreezed() 15`] = `false`;

exports[`Tag Object > isFreezed() 16`] = `false`;

exports[`Tag Object > isFreezed() 17`] = `true`;

exports[`Tag Object > isFreezed() 18`] = `false`;

exports[`Tag Object > isFreezed() 19`] = `false`;

exports[`Tag Object > isFreezed() 20`] = `true`;

exports[`Tag Object > isFreezed() 21`] = `false`;

exports[`Tag Object > isFreezed() 22`] = `false`;

exports[`Tag Object > isFreezed() 23`] = `false`;

exports[`Tag Object > isFreezed() 24`] = `false`;
